packages/@aws-cdk/aws-codedeploy/lib/deployment-group.ts

```diff
@@ -142,18 +171,80 @@ export class ServerDeploymentGroup extends ServerDeploymentGroupRef {
             managedPolicyArns: ['arn:aws:iam::aws:policy/service-role/AWSCodeDeployRole'],
         });
 
+        this._autoScalingGroups = props.autoScalingGroups || [];
+        this.autoInstallAgent = props.autoInstallAgent !== false;
+        this.codeDeployBucket = s3.BucketRef.import(this, 'CodeDeployBucket', {
+            bucketName: `aws-codedeploy-${this.region}`,
+        });
+        for (const asg of this._autoScalingGroups) {
+            this.addCodeDeployAgentInstallUserData(asg);
+        }
+
         const resource = new cloudformation.DeploymentGroupResource(this, 'Resource', {
             applicationName: this.application.applicationName,
             deploymentGroupName: props.deploymentGroupName,
             serviceRoleArn: this.role.roleArn,
             deploymentConfigName: props.deploymentConfig &&
                 props.deploymentConfig.deploymentConfigName,
+            autoScalingGroups: new cdk.Token(() =>
+                this._autoScalingGroups.length === 0
+                    ? undefined
+                    : this._autoScalingGroups.map(asg => asg.autoScalingGroupName())),
         });
 
         this.deploymentGroupName = resource.ref.toString();
         this.deploymentGroupArn = deploymentGroupName2Arn(this.application.applicationName,
             this.deploymentGroupName);
     }
+
+    public addAutoScalingGroup(asg: autoscaling.AutoScalingGroup): void {
+        this._autoScalingGroups.push(asg);
+        this.addCodeDeployAgentInstallUserData(asg);
+    }
+
+    public get autoScalingGroups(): autoscaling.AutoScalingGroup[] {
+        return this._autoScalingGroups.slice();
+    }
+
+    private addCodeDeployAgentInstallUserData(asg: autoscaling.AutoScalingGroup): void {
+        if (!this.autoInstallAgent) {
+            return;
+        }
+
+        this.codeDeployBucket.grantRead(asg.role, 'latest/install');
+
+        switch (asg.osType) {
+            case ec2.OperatingSystemType.Linux:
+                asg.addUserData(
```

Comment:
Assuming this user data has been fully tested on different OS per https://docs.aws.amazon.com/codedeploy/latest/userguide/codedeploy-agent.html